org.adempiere.base/src/org/compiere/process/ProcessInfo.java

```diff
@@ -691,6 +693,139 @@ public ProcessInfoLog[] getLogs()
+	{
+		return saveLog (new ProcessInfoLog (Log_ID, P_ID, P_Date, P_Number, P_Msg));
+	}	//	saveLog
+
+	/**
+	 * 	Save Log to DB immediately
+	 *	@param P_ID Process ID
+	 *	@param P_Date Process Date
+	 *	@param P_Number Process Number
+	 *	@param P_Msg Process Message
+	 *	@return String AD_PInstance_Log_UU
+	 */
+	public String saveLog (int P_ID, Timestamp P_Date, BigDecimal P_Number, String P_Msg)
+	{
+		return saveLog (new ProcessInfoLog (P_ID, P_Date, P_Number, P_Msg));
+	}	//	saveLog
+
+	/**
+	 * 	Save Log to DB immediately
+	 *	@param logEntry log entry
+	 *	@return String AD_PInstance_Log_UU
+	 */
+	public String saveLog (ProcessInfoLog logEntry)
+	{
+		if (logEntry == null)
+			return "";
+		MPInstanceLog il = new MPInstanceLog(getAD_PInstance_ID(), 
+				logEntry.getLog_ID(), 
+				logEntry.getP_Date(),
+				logEntry.getP_ID(), 
+				logEntry.getP_Number(), 
+				logEntry.getP_Msg(),
+				logEntry.getAD_Table_ID(), 
+				logEntry.getRecord_ID(),
+				!Util.isEmpty(logEntry.getPInstanceLogType()) ? logEntry.getPInstanceLogType() : X_AD_PInstance_Log.PINSTANCELOGTYPE_Progress);
+		il.saveEx();
+		return il.getAD_PInstance_Log_UU();
+	}	//	saveLog
+	
+	/**
+	 * Update Log to DB immediately
+	 * @param pInstanceLogUU AD_PInstance_Log_UU
+	 * @param P_ID Process ID
+	 * @param P_Date Process Date
+	 * @param P_Number Process Number
+	 * @param P_Msg Process Message
+	 * @param tableId AD_Table_ID
+	 * @param recordId Record_ID
+	 * @return true if log is successfully updated
+	 */
+	public boolean updateLog (String pInstanceLogUU, int P_ID, Timestamp P_Date, BigDecimal P_Number, String P_Msg,int tableId,int recordId)
+	{
+		return updateLog (new ProcessInfoLog (pInstanceLogUU, P_ID, P_Date, P_Number, P_Msg,tableId,recordId));
+	}	//	updateLog
+	
+	/**
+	 * 	Update Log to DB immediately
+	 *	@param pInstanceLogUU AD_PInstance_Log_UU
+	 *	@param P_ID Process ID
+	 *	@param P_Date Process Date
+	 *	@param P_Number Process Number
+	 *	@param P_Msg Process Message
+	 *	@return true if log is successfully updated
+	 */
+	public boolean updateLog (String pInstanceLogUU, int P_ID, Timestamp P_Date, BigDecimal P_Number, String P_Msg)
+	{
+		return updateLog (new ProcessInfoLog (pInstanceLogUU, P_ID, P_Date, P_Number, P_Msg));
+	}	//	updateLog
+	
+	/**
+	 * 	Update existing Log immediately
+	 *	@param logEntry log entry
+	 *	@return true if log is successfully updated
+	 */
+	public boolean updateLog (ProcessInfoLog logEntry)
+	{
+		if (logEntry == null)
+			return false;
+		MPInstanceLog il = new MPInstanceLog(logEntry.getAD_PInstance_Log_UU(),
+				getAD_PInstance_ID(), 
+				logEntry.getLog_ID(), 
+				logEntry.getP_Date(),
+				logEntry.getP_ID(), 
+				logEntry.getP_Number(), 
+				logEntry.getP_Msg(),
+				logEntry.getAD_Table_ID(), 
+				logEntry.getRecord_ID(),
+				!Util.isEmpty(logEntry.getPInstanceLogType()) ? logEntry.getPInstanceLogType() : X_AD_PInstance_Log.PINSTANCELOGTYPE_Status);
+		return il.update();
+	}	//	saveLog
+	

```

Comment:
The multiple savelog and updatelog method here is heavily bias towards the log type of "P", would prefer a more balance API.

I'm not sure what to make up of the new methods added here as it seems you are using the "P"rogress log type in a very specific way but the API added here doesn't reflect that and there isn't any examples provided.

Reply:
hi @PeterTakacs300 , I think it is easier and more intuitive for developer to have "logProgress", "updateProgress" and "logStatus" method instead. Also, unless there's real need of it, not to have multiple variants of "logProgress" and "logStatus" method to keep the API simple.

WDYT?

Reply:
Hi @hengsin, I've made a new commit to the Generate Random Data for SO/PO process here: https://github.com/cloudempiere/org.cloudempiere.generaterandomorders/pull/2
There is an example of the use of the saveLog and updateLog methods.

The initial purpose of this PR was to add the possibility to create process audit logs while the process is running. I thought it could be quite intuitive to use save and update log methods (saveLog should always create a new log, updateLog updates existing log), and the log type is assigned automatically, so the implementor of the process should not worry about it.

Do you think it would be more balanced, if we would give setting the log type in the hands of the process implementor (by adding Log Type argument to the methods)?

Reply:
That's exactly what I means above, the method is design with just the "P" log type in mind. I've already posted my preference above (i.e replace saveLog and updateLog with logProgress, updateProgress and logStatus, the method name should be self explanatory here). However, if you prefer to just add a log type argument, that should be fine too.

Reply:
hi @PeterTakacs300 , what you have just committed is different from your message above. Please clarify.

Reply:
hi @hengsin, maybe there is a mismatch in the log type names in what you mentioned. If I'm correct it should be like this:
- Status log - displays always the current status - so this is the log type that can be updated
- Progress log - we originally named it 'Tail', changes to Progress after Chuck's suggestion - these are single logs saved continuously through the execution of the process
- Result log - legacy

Only Status logs can be updated.

In SvrProcess I renamed the methods as you suggested, but in the ProcessInfo layer I stayed with the saveLog/updateLog naming and added log type argument (to reduce the number of methods here, since the process developers usually do not use the logging methods from this level anyways). Can it stay like this, or ProcessInfo method naming should match the ones in SvrProcess?

Reply:
I think it is simpler to comprehend to have similar named method between the 2 class.

For me, Status and Progress should be the other way round, I think it is more common to say "Show Progress of a Process" (i.e 30%, 50%, etc).

Reply:
hi @hengsin, methods are renamed and logic exchanged between Status and Progress